nodeJS/authentication/sessions.md

```diff
@@ -0,0 +1,463 @@
+
+### Implementing sessions
+
+Let's use [express-session](https://expressjs.com/en/resources/middleware/session.html) to implement a basic session authentication system - code along! For the purpose of streamlining our example, we'll put all of the JavaScript in `app.js` and start with hardcoding a few things. When it comes to your own projects, separate different parts like routes, controllers etc. to their own files and folders as you'll have done before. The same can be done for any configuration for authentication features.
+
+#### Setup
+
+First, make a new database within `psql`. We will start by setting up a `users` table:
+
+```sql
+CREATE TABLE users (
+  id INTEGER PRIMARY KEY GENERATED ALWAYS AS IDENTITY,
+  username VARCHAR ( 255 ),
+  password VARCHAR ( 255 )
+);
+```
+
+Now we'll set up a minimal Express app. We'll need the following dependencies installed first:
+
+```bash
+npm install express express-session ejs dotenv pg connect-pg-simple
+```
+
+Now our initial `index.ejs` view:
+
+```ejs
+<!-- views/index.ejs -->
+<!DOCTYPE html>
+<html lang="en">
+<head>
+  <meta charset="UTF-8">
+  <title>Home</title>
+</head>
+<body>
+  <h1>Hello world!</h1>
+</body>
+</html>
+```
+
+And our `app.js`:
+
+```javascript
+// app.js
+require("dotenv").config();
+const path = require("node:path");
+const { Pool } = require("pg");
+const express = require("express");
+const session = require("express-session");
+const pgSession = require("connect-pg-simple")(session);
+
+const pool = new Pool({
+  // add your db configuration
+});
+
+const app = express();
+app.set("views", path.join(__dirname, "views"));
+app.set("view engine", "ejs");
+
+app.use(session({
+  store: new pgSession({
+    pool: pool,
+    createTableIfMissing: true,
+  }),
+  resave: false,
+  saveUninitialized: false,
+  secret: process.env.SESSION_SECRET,
+  cookie: {
+    httpOnly: process.env.NODE_ENV === 'prod',
+    secure: process.env.NODE_ENV === 'prod',
+    maxAge: 2 * 24 * 60 * 60 * 1000, // 2 days
+  },
+}));
+app.use(express.urlencoded({ extended: false }));
+
+app.get("/", (req, res) => {
+  res.render("index");
+});
+
+const PORT = process.env.PORT || 3000;
+app.listen(PORT, () => {
+  console.log(`App listening on port ${PORT}!`);
+});
+```
+
+#### Session store
+
+Let's talk about our session config which we apply to every incoming request (by mounting it on `app`). Firstly, we use the [connect-pg-simple](https://www.npmjs.com/package/connect-pg-simple) library to make express-session store session data in a "session" table in our database, creating the table if it does not already exist. If you look inside your database in psql, you'll be able to see what the session table looks like. You can also have a look at the [SQL queries for creating the session table](https://github.com/voxpelli/node-connect-pg-simple/blob/HEAD/table.sql).
```

Comment:
I think we should ask learners to look at the table.

```suggestion
Let's talk about our session config which we apply to every incoming request (by mounting it on `app`). Firstly, we use the [connect-pg-simple](https://www.npmjs.com/package/connect-pg-simple) library to make express-session store session data in a "session" table in our database, creating the table if it does not already exist. Go inside your database in psql, you'll be able to see what the session table looks like. You can also have a look at the [SQL queries for creating the session table](https://github.com/voxpelli/node-connect-pg-simple/blob/HEAD/table.sql).
```

I think we should explain the session table ourselves as well. Just some text on each column's purpose should do